vermouth/processors/do_links.py

```diff
@@ -28,26 +29,79 @@
 
 
 class LinkGraphMatcher(nx.isomorphism.GraphMatcher):
-    def semantic_feasibility(self, node1, node2):
+    def semantic_feasibility(self, node1_name, node2_name):
         # TODO: implement (partial) wildcards
         # Node2 is the link
-        node1 = self.G1.nodes[node1]
-        node2 = self.G2.nodes[node2]
-        for attr in node2:
-            if attr == 'order':
-                continue
-            if node1.get(attr, None) != node2[attr]:
+        node1 = self.G1.nodes[node1_name]
+        node2 = self.G2.nodes[node2_name]
+        return _atoms_match(node1, node2)
+
+
+def _atoms_match(node1, node2):
+    for attr in node2:
+        if attr in ['order', 'replace']:
+            continue
+        if isinstance(node2[attr], Choice):
+            if node1.get(attr, None) not in node2[attr]:
+                return False
+        elif node1.get(attr, None) != node2[attr]:
+            return False
+    else:
+        return True
+
+
+def _is_valid_non_edges(molecule, link, rev_raw_match):
+    for from_node, to_node_attrs in link.non_edges:
+        if from_node not in link:
+            continue
+        from_mol_node_name = rev_raw_match[from_node]
+        from_mol = molecule.nodes[from_mol_node_name]
+        from_link = link.nodes[from_node]
+        from_resid = from_mol['resid']
+        from_order = from_link.get('order', 0)
+        for neighbor in molecule.neighbors(from_mol_node_name):
+            to_mol = molecule.nodes[neighbor]
+            to_link = to_node_attrs
+            to_resid = to_mol['resid']
+            to_order = to_link.get('order', 0)
+            if to_resid == from_resid + to_order and _atoms_match(to_mol, to_link):
                 return False
-        else:
-            return True
+    return True
+
 
+def _pattern_match(molecule, atoms, raw_match):
+    for link_key, template_attr in atoms:
+        try:
+            molecule_key = raw_match[link_key]
+        except KeyError as err:
+            print(raw_match)
```

Comment:
Prettier message?

Reply:
This was there for debug purpose. There is no reason to catch the exception. Removed.